### PR TITLE
[cxl-notification]

### PR DESCRIPTION
**Pull request by anoblet**
https://app.clickup.com/t/3d9ttgy

### CODE REVIEW

**Comment by saas786** [comment]
@anoblet 

- test & squash commits in [this](https://github.com/conversionxl/aybolit/pull/218) PR
- Merge: https://github.com/conversionxl/aybolit/pull/219 into [this](https://github.com/conversionxl/aybolit/pull/218) PR
- also rebase it from `master` branch.

I will review & test it one last time.

Looks good for now.

**Review by saas786 — APPROVED** [review]

**Review by pawelkmpt — COMMENTED** [review]
I looked at netlify preview and noticed that `x` is not aligned correctly in close button. I checked new notification but also GF.

Screenshot 2022-09-22 at 05 47 49
Screenshot 2022-09-22 at 05 48 06
Screenshot 2022-09-22 at 05 48 15

**Comment by saas786** [comment]
> I looked at netlify preview and noticed that `x` is not aligned correctly in close button. I checked new notification but also GF.
> 
> Screenshot 2022-09-22 at 05 47 49
> 
> Screenshot 2022-09-22 at 05 48 06
> 
> Screenshot 2022-09-22 at 05 48 15

@pawelkmpt its not a part of `cxl-notification` itself, but rather focused on Gravity Forms styles...

You can see in screenshot below, that `margin-top` is applied, also it could use smaller font size as well.

chrome_dbothXKzSi

**Comment by saas786** [comment]
Issue created for Gravity Forms styles in notification: https://github.com/conversionxl/aybolit/issues/225

**Comment by pawelkmpt** [comment]
> @pawelkmpt its not a part of cxl-notification itself, but rather focused on Gravity Forms styles...

But this part is about notification. `x` inside the red box is not in the middle.

> I looked at netlify preview and noticed that `x` is not aligned correctly in close button. I checked new notification but also GF.
> 
> Screenshot 2022-09-22 at 05 47 49


> You can see in screenshot below, that margin-top is applied, also it could use smaller font size as well.

I don't think top margin should have impact on centering `x` inside the button.

**Comment by saas786** [comment]
> > @pawelkmpt its not a part of cxl-notification itself, but rather focused on Gravity Forms styles...
> 
> But this part is about notification. `x` inside the red box is not in the middle.
> 
> > I looked at netlify preview and noticed that `x` is not aligned correctly in close button. I checked new notification but also GF.
> > Screenshot 2022-09-22 at 05 47 49
> 
> > You can see in screenshot below, that margin-top is applied, also it could use smaller font size as well.
> 
> I don't think top margin should have impact on centering `x` inside the button.

Close icon is usually at top right, not middle aligned.

We can center align it, but it would only look nice for single line notifications, multiline it would not good aligned or good.
chrome_io3kKYHykq

**Comment by pawelkmpt** [comment]
> Close icon is usually at top right, not middle aligned.
> 
> We can center align it, but it would only look nice for single line notifications, multiline it would not good aligned or good.

You don't get what I mean by _"`x` inside the red box is not in the middle"_.

**BAD**

Screenshot 2022-09-22 at 14 03 51


**GOOD**
Screenshot 2022-09-22 at 14 04 40

**Comment by saas786** [comment]
> > Close icon is usually at top right, not middle aligned.
> > We can center align it, but it would only look nice for single line notifications, multiline it would not good aligned or good.
> 
> You don't get what I mean by _"`x` inside the red box is not in the middle"_.
> 
> **BAD**
> 
> Screenshot 2022-09-22 at 14 03 51
> 
> **GOOD** Screenshot 2022-09-22 at 14 04 40

Correct, I didn't understood it :), thanks for re-explaining.

But it still only applicable to GravityForms styles...

See un-paid order notification button, centered.

chrome_tIEVYklaQu

See Gravity Forms notification, and un-paid notifcation buttons, but have same styles applied, only calculated width is different, which is causing it to look mis-aligned.


chrome_YDaUqFatR4

<hr/>

chrome_YykNrP3qHr

So can delegate it to https://github.com/conversionxl/aybolit/issues/225

**Comment by pawelkmpt** [comment]
> Correct, I didn't understood it :), thanks for re-explaining.
> 
> But it still only applicable to GravityForms styles...
> 
> See un-paid order notification button, centered.

Issue is present for notification as well on mobile screens. I should've mentioned that in the beginning.
Tablet and desktop indeed look good.

> So can delegate it to https://github.com/conversionxl/aybolit/issues/225

It has to be done here.

Screenshot 2022-09-28 at 10 04 13


Screenshot 2022-09-28 at 10 03 58

**Comment by anoblet** [comment]
I found the culprit. This `h2` is stretching the button somehow: https://github.com/conversionxl/aybolit/blob/6de178f17450a6a2e868b9c8d994cea14943e4fb/packages/storybook/cxl-ui/cxl-notification/gravity-form.story.js#L17-L20

If you replace this with a simple `div`, the icon is centered.

**Comment by anoblet** [comment]
This should be fixed by wrapping the button in a `div`.

**Comment by lkraav** [comment]
> I found the culprit. This h2 is stretching the button somehow

Why would the button be *inside* heading? Super weird structure.

**Comment by anoblet** [comment]
@lkraav 

For the default story, I took the structure from the Vaadin demo (https://vaadin.com/docs/latest/components/notification/#error). Vaadin doesn't recommend long content for notifications (https://vaadin.com/docs/latest/components/notification/#limit-content-length)

**Comment by saas786** [comment]
> > I found the culprit. This h2 is stretching the button somehow
> 
> Why would the button be _inside_ heading? Super weird structure.

Its not inside `<h2>` but rather alongside it, as sibling.

**Comment by saas786** [comment]
> This should be fixed by wrapping the button in a div.

@anoblet 

Patched the `<vaadin-button>`, I added it to global styles, couldn't find proper selector to make it specific for `<cxl-notification>`, can you please look into making it so it applies this style when `<cxl-notification><vaadin-button>..` its wrapped inside `<cxl-notification>`.

Inspired from:

firefox_jV57o3dm6p

**Comment by anoblet** [comment]
@saas786 I had it fixed in the commit I pushed.

I understand the `h2` is a sibling of `vaadin-button`. The `h2` stretches the content so that it spans multiple lines. This affects the dimensions of the button. The reason the default story doesn't have this issue is because it only spans one line.

There is no CSS necessary here. Just wrap the button in a `div`.

```
<vaadin-notification>
  <h2>...</h2>
  <div>
    <vaadin-button>...</vaadin-button>
  </div>
</vaadin-notification>
```

We should also be using `vaadin-icon` as opposed to `iron-icon` as `iron-icon` is deprecated.

**Comment by saas786** [comment]
> @saas786 I had it fixed in the commit I pushed.
> 
> I understand the `h2` is a sibling of `vaadin-button`. The `h2` stretches the content so that it spans multiple lines. This affects the dimensions of the button. The reason the default story doesn't have this issue is because it only spans one line. Multiple line notifications are discouraged by Vaadin ([vaadin.com/docs/latest/components/notification/#limit-content-length](https://vaadin.com/docs/latest/components/notification/#limit-content-length)) and should probably be replaced with a dialog.
> 
> If you really want a multiple line notification, there is no CSS necessary. Wrap the button in a `div`.
> 
> ```
> <vaadin-notification>
>   <h2>...</h2>
>   <div>
>     <vaadin-button>...</vaadin-button>
>   </div>
> </vaadin-notification>
> ```
> 

I understood your reasoning, and it was working as well, 
but markup change is not the solution here, so I discarded it. 
Please work on css change I pushed to branch, and make it Shadow Dom specific selector for `<cxl-notification>` so `vaadin-notification-card` should be target.

> We should also be using `vaadin-icon` not `iron-icon`. `iron-icon` is deprecated. This was also fixed in my commit.

 `iron-icon` is from live website, its not hard code, 
 so once we swap out `iron-icon` with `vaadin-icon` we can update this story.

**Comment by anoblet** [comment]
Vaadin appends the notification to the document root. The closest rule/selector we can get is:
```
vaadin-notification-card vaadin-button[theme~="icon"]:not([theme~="tertiary-inline"]) {
  flex: none;
}
```

**Comment by pawelkmpt** [comment]
@anoblet can we squash a895dd266f26c9fcd49945232c18256ee6ad2918 and 7420f4489120fdebc6f1bb227f1676c734c02498 ?

**Review by pawelkmpt — APPROVED** [review]

File: packages/cxl-ui/src/components/cxl-notification.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  renderer(root) {
+    if (root.firstElementChild) {
+      return;
+    }
+
+    [...this.childNodes].forEach((node) => {
+      root.appendChild(node);
+    });
+  }
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-notification.js:16]: This could've used a a comment for "why is this necessary" @anoblet

Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-notification.js:16]: Makes sense. I will file a PR to add the comment.
